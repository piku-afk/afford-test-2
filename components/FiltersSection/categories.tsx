import { Stack } from '@mui/material';
import { Box } from '@mui/system';
import { FC } from 'react';
import { CategoryCheckBox, StyledLabel } from '../StyledMuiComponents';
import { useCategories } from '@/hooks/useCategories';

type CategoriesProps = {
  value: {
    [key: string]: {
      [key: string]: boolean;
    };
  };
  handleChange: (params: {
    parent: string;
    values: { [key: string]: boolean };
  }) => void;
};

export const Categories: FC<CategoriesProps> = (props) => {
  const { value, handleChange } = props;
  const { categories, handleCheckAll, handleCheckBoxChange } =
    useCategories(handleChange);

  return (
    <Stack sx={{ paddingLeft: 1 }} spacing={1}>
      {categories.map((category) => {
        const { ID, name: parentCategory, subCategories } = category;
        const values = Object.values(value[parentCategory] || {}) || [];
        const parentCheck = values.length > 0 && values.some((value) => value);

        return (
          <Box key={ID}>
            <StyledLabel
              label={parentCategory}
              sx={{
                '& .MuiFormControlLabel-label': {
                  fontWeight: parentCheck ? 600 : 'inherit',
                },
              }}
              control={
                <CategoryCheckBox
                  size='small'
                  color='secondary'
                  checked={parentCheck}
                  onChange={(e) =>
                    handleCheckAll(
                      e.target.checked,
                      parentCategory,
                      subCategories.map(({ name }) => name)
                    )
                  }
                />
              }
            />
            <Stack sx={{ ml: 3 }}>
              {subCategories.map((category) => {
                const { ID, name } = category;
                const { [name]: checked = false } = value[parentCategory] || {};

                return (
                  <StyledLabel
                    key={ID}
                    label={name}
                    control={
                      <CategoryCheckBox
                        size='small'
                        color='secondary'
                        checked={checked}
                        onChange={(e) =>
                          handleCheckBoxChange(
                            name,
                            Boolean(e.target.value),
                            parentCategory
                          )
                        }
                      />
                    }
                  />
                );
              })}
            </Stack>
          </Box>
        );
      })}
    </Stack>
  );
};
