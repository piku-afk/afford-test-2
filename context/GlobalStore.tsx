import {
  Dispatch,
  FC,
  PropsWithChildren,
  useContext,
  useReducer,
  createContext,
  useEffect,
} from 'react';
import { initialState } from './initialState';
import { ActionInterface, ActionTypes, reducer } from './reducer';

interface GlobalContextType {
  state: typeof initialState;
  dispatch: Dispatch<ActionInterface>;
}

export const GlobalContext = createContext<GlobalContextType>({
  state: initialState,
  dispatch: () => initialState,
});

export const useGlobalStore = () => useContext(GlobalContext);

type GlobalStoreProps = {};

export const GlobalStore: FC<PropsWithChildren<GlobalStoreProps>> = (props) => {
  const { children } = props;
  const [state, dispatch] = useReducer(reducer, initialState);

  useEffect(() => {
    const init = async () => {
      const res = await fetch('http://localhost:5000/filters');
      const data = await res.json();
      const { searchFilter } = data || {};
      const { brands, categories } = searchFilter || {};

      dispatch({ type: ActionTypes.setCategories, payload: categories });
      dispatch({ type: ActionTypes.setBrands, payload: brands });
    };
    init();
  }, []);

  const value = { state, dispatch };

  return (
    <GlobalContext.Provider value={value}>{children}</GlobalContext.Provider>
  );
};
