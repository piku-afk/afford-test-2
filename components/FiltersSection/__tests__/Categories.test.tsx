import * as GlobalContext from '@/context/GlobalStore';
import { Categories } from '../categories';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

const user = userEvent.setup();
const mockCategories = [
  {
    ID: '1',
    name: 'hello',
    subCategories: [{ ID: '2', name: 'world', subCategories: [] }],
  },
  {
    ID: '2',
    name: 'second category',
    subCategories: [{ ID: '2', name: 'sub cat', subCategories: [] }],
  },
];

const handleChange = jest.fn();

const MockCategory = () => {
  return (
    <Categories
      value={{ hello: { world: true } }}
      handleChange={handleChange}
    />
  );
};

describe('Cart', () => {
  beforeEach(() => {
    jest.spyOn(GlobalContext, 'useGlobalStore').mockReturnValue({
      state: { brands: [], categories: mockCategories, cart: [] },
      dispatch: jest.fn,
    });
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<MockCategory />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders the checkboxes correctly', () => {
    render(<MockCategory />);

    const helloCheckBox = screen.getByRole('checkbox', { name: /hello/i });
    expect(helloCheckBox).toBeInTheDocument();
    expect(helloCheckBox).toBeChecked();

    const worldCheckBox = screen.getByRole('checkbox', {
      name: /world/i,
    });
    expect(worldCheckBox).toBeInTheDocument();
    expect(worldCheckBox).toBeChecked();

    const secondCheckBox = screen.getByRole('checkbox', {
      name: /second category/i,
    });
    expect(secondCheckBox).toBeInTheDocument();
    expect(secondCheckBox).not.toBeChecked();

    const subCatCheckBox = screen.getByRole('checkbox', {
      name: /sub cat/i,
    });
    expect(subCatCheckBox).toBeInTheDocument();
    expect(subCatCheckBox).not.toBeChecked();
  });

  it('renders the checkboxes correctly when clicked', async () => {
    render(<MockCategory />);

    const helloCheckBox = screen.getByRole('checkbox', { name: /hello/i });
    const worldCheckBox = screen.getByRole('checkbox', {
      name: /world/i,
    });
    const subCatCheckBox = screen.getByRole('checkbox', {
      name: /sub cat/i,
    });

    await user.click(helloCheckBox);
    await user.click(subCatCheckBox);
    await user.click(worldCheckBox);

    expect(handleChange).toHaveBeenCalled();
    expect(handleChange).toHaveBeenCalledTimes(3);
  });
});

export {};
