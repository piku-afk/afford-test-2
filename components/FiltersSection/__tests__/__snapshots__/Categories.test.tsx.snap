// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-14cqa6r-MuiStack-root"
  >
    <div
      class="MuiBox-root css-0"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1inlnfb-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root Mui-checked css-1wo3ca8-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="CheckBoxIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          hello
        </span>
      </label>
      <div
        class="css-tq04rj-MuiStack-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1i4bd1i-MuiFormControlLabel-root"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root Mui-checked css-1wo3ca8-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              checked=""
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CheckBoxIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            world
          </span>
        </label>
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <label
        class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1mmh5c2-MuiFormControlLabel-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root css-1wo3ca8-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="CheckBoxOutlineBlankIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
        <span
          class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
        >
          second category
        </span>
      </label>
      <div
        class="css-tq04rj-MuiStack-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-1i4bd1i-MuiFormControlLabel-root"
        >
          <span
            class="MuiCheckbox-root MuiCheckbox-colorSecondary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorSecondary PrivateSwitchBase-root css-1wo3ca8-MuiButtonBase-root-MuiCheckbox-root"
          >
            <input
              class="PrivateSwitchBase-input css-1m9pwf3"
              data-indeterminate="false"
              type="checkbox"
            />
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
              data-testid="CheckBoxOutlineBlankIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            sub cat
          </span>
        </label>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
